src/octoprint/static/js/app/helpers.js

```diff
@@ -1619,3 +1623,80 @@ var deepMerge = function (target, source) {
 
     return target;
 };
+
+var isWebRTCAvailable = function () {
+    return (
+        typeof RTCPeerConnection == "function" &&
+        typeof RTCPeerConnection.prototype.addEventListener == "function" &&
+        typeof RTCPeerConnection.prototype.addTransceiver == "function" &&
+        typeof RTCPeerConnection.prototype.createOffer == "function" &&
+        typeof RTCPeerConnection.prototype.setLocalDescription == "function" &&
+        typeof RTCPeerConnection.prototype.removeEventListener == "function" &&
+        typeof RTCPeerConnection.prototype.addEventListener == "function" &&
+        typeof RTCPeerConnection.prototype.setRemoteDescription == "function"
+    );
+};
+
+var negotiateWebRTC = function (streamUrl) {
+    pc.addTransceiver("video", {direction: "recvonly"});
+    pc.addTransceiver("audio", {direction: "recvonly"});
+    return pc
+        .createOffer()
+        .then(function (offer) {
+            return pc.setLocalDescription(offer);
+        })
+        .then(function () {
+            // Wait for ICE gathering to complete
+            return new Promise(function (resolve) {
+                if (pc.iceGatheringState === "complete") {
+                    resolve();
+                } else {
+                    function checkState() {
+                        if (pc.iceGatheringState === "complete") {
+                            pc.removeEventListener("icegatheringstatechange", checkState);
+                            resolve();
+                        }
+                    }
+                    pc.addEventListener("icegatheringstatechange", checkState);
+                }
+            });
+        })
+        .then(function () {
+            var offer = pc.localDescription;
+            // webrtc://host.com becomes http://host.com
+            // webrtcs://host.com becomes https://host.com
+            streamUrl = "http" + streamUrl.slice("webrtc".length);
+            return $.ajax({
+                url: streamUrl,
+                type: "POST",
+                dataType: "json",
+                data: JSON.stringify({
+                    sdp: offer.sdp,
+                    type: offer.type
+                }),
+                contentType: "application/json; charset=UTF-8"
+            });
+        })
+        .then(function (response) {
+            return pc.setRemoteDescription(response);
+        })
+        .catch(function (e) {
+            console.error(e);
+        });
+};
+
+var startWebRTC = function (videoElement, streamUrl) {
+    var config = {
+        sdpSemantics: "unified-plan",
+        iceServers: [{urls: ["stun:stun.l.google.com:19302"]}]
```

Comment:
Hm... should we maybe turn this into an (advanced) option? There are people out there who prefer no contact with google services of any kind, so allowing to specific the stun server to use would be a good idea.

Reply:
I've pushed a commit to address this, please check on your end as well.

I tested this based on the instructions from your OP and the feed works. However, I'm seeing an interesting issue where it only works *once*. I have to restart the server side between page loads or all that I get is a blank video. No idea if that is an issue on my end or with the server. Tested with a C270 and the transcoding command line.

Reply:
Good call thanks for handling that!

It's almost certainly a problem with the server, that thing is still rough around the edges. Did you pull the latest from my `aiortc` `octoprint-webrtc-support` branch? I made a few tweaks on Sunday. Anything interesting in stdout from the server?

The good thing about the front end code in this PR is that it should be pretty agnostic to server implementation so we should be able to swap in any WebRTC server if `aiortc` doesn't fit our needs (Janus for example is the incumbent option).

Reply:
Nothing on stdout, looks just like it should work. And this was a clone from this morning. Eh 🤷‍♀️

Reply:
🤷 interesting. You can always add a `-vv` flag to the `webcam.py` command to get verbose output. But I think fine to punt on it for now -- I'll test your commit here on my end as well.